IPNI_FEDERATION.md

```diff
@@ -0,0 +1,425 @@
+
+However, the realization of the vision of IPNI was not without its challenges. As the adoption grew and instances
+sprouted globally, the heterogeneity in content discovery, attributed to the variability in provider lists across
+instances, became evident. A singular instance, despite its comprehensive network state, could only discover content at
+a pace and breadth defined by its provider list.
+
+Furthermore, the rise of content-centric networks, where data is retrieved based on content rather than its location,
+emphasized the importance of a unified and consistent view. The fragmented state of index records across instances
+threatened the core utility of IPNI as a reliable and quick lookup mechanism.
+
+The necessity for a federation protocol, therefore, became paramount. It was envisioned not just as a bridge to connect
+multiple IPNI instances but as a robust mechanism to ensure that the distributed nature of the system did not compromise
+on the uniformity and reliability of content discovery. This document delves deep into this protocol, exploring its
+intricacies and mechanisms, laying the foundation for a truly inter-planetary scale network indexer.
+
+## Motivation
+
+The underlying motivation behind enhancing IPNI is manifold. Foremost, guaranteeing an eventually consistent view of
+index records across multiple IPNI instances holds the potential for transformative benefits. Such consistency paves the
+way for reducing trust in an individual indexer.
+
+## Non-goals
+
+This section clarifies aspects that the IPNI federation protocol will not address in its current iteration.
+Here are the non-goals for the IPNI federation protocol:
+
+* **Permissionless Membership**: This specification assumes a permissioned network, where the membership is controlled
+  and not open to arbitrary participants.
+* **Real-time synchronization**: The protocol strives for eventual consistency, but real-time synchronization across all
+  nodes, especially in a globally distributed setup, is not the objective.
+* **Network Topology Design**: The IPNI federation protocol does not define or restrict the underlying network topology.
+  Whether nodes are structured in a mesh, star, or hierarchical format, the protocol remains agnostic.
+* **Strong consistency**: While the IPNI federation protocol aims to ensure data consistency and availability, it does
+  not aim to atomically replicate every piece of indexed data across all nodes. For example, it is possible for some
+  records to remain partially available on one node until deletions are propagate through the network.
+
+## Terminology
+
+* **Epoch**: An epoch in the context of the IPNI federation protocol refers to a specific time period or iteration of
+  the protocol's operation. It can be thought of as a version or generation number that increments each time there's a
+  major operation or update, aiding in synchronization efforts across instances.
```

Comment:
Who determines what an epoch consists of? Different IPNI instances may have different rates of progress for different index providers, depending on network location, bandwidth limits, etc. If any single IPNI instance tries to declare an epoch based on its view of indexing progress, then other IPNI instance may never be able to reach consistency because they will always be ahead for some index providers.

I think the means an epoch is a set of agreed-on progress points for all index providers in the federation, that each IPNI node can independently determine. A distance from the root of each provider's advertisement chain could serve this purpose. A date may not be a good example unless the IPNI nodes have access to a shared clock.

Reply:
this sounds like consensus on a vector clock?

Reply:
Different epochs can contain different sets of index providers, but an epoch has a specific set of index providers associated with it.